main.go

```diff
@@ -106,22 +124,109 @@ func main() {
 			labelValue:             labelValue,
 		}
 		c := newController(klient, logger, opt)
+		c.registerMetrics(reg)
 		done := make(chan struct{})
 
 		g.Add(func() error {
 			return c.run(done)
-		}, func(err error) {
+		}, func(_ error) {
 			level.Info(logger).Log("msg", "shutting down controller")
 			close(done)
 		})
 	}
-	level.Info(logger).Log("msg", "starting the controller")
+	{
+		router := http.NewServeMux()
+		router.Handle("/metrics", promhttp.InstrumentMetricHandler(reg, promhttp.HandlerFor(reg, promhttp.HandlerOpts{})))
+		srv := &http.Server{Addr: config.InternalAddr, Handler: router}
 
+		g.Add(func() error {
+			return srv.ListenAndServe()
+		}, func(err error) {
+			if err != http.ErrServerClosed {
+				level.Warn(logger).Log("msg", "internal server closed unexpectedly")
+				return
+			}
+			level.Info(logger).Log("msg", "shutting down internal server")
+			ctx, _ := context.WithTimeout(context.Background(), internalServerShutdownTimeout)
+			if err := srv.Shutdown(ctx); err != nil {
+				stdlog.Fatal(err)
+			}
+		})
+	}
+
+	level.Info(logger).Log("msg", "starting the controller")
 	if err := g.Run(); err != nil {
 		stdlog.Fatal(err)
 	}
 }
 
+type prometheusReflectorMetrics struct {
+	reg *prometheus.Registry
+}
+
+func newReflactorMetrics(reg *prometheus.Registry) prometheusReflectorMetrics {
+	return prometheusReflectorMetrics{reg}
+}
+
+func (p prometheusReflectorMetrics) counter(name string) prometheus.Counter {
+	counter := prometheus.NewCounter(prometheus.CounterOpts{
+		Subsystem: "thanos-receive-controller",
```

Comment:
metrics should have underscores instead of dashes, and `.Subsystem` is targeted for removal, so just put it directly in the name